src/cpp/_pythia8.cpp

```diff
@@ -103,6 +105,73 @@ py::array_t<int> event_array_children(Event &event)
     return py::array_t<int>(shape, strides, ptr1);
 }
 
+// refills "event" stack with particles
+void fill_event(Event &event,
+                    py::array_t<int> &pid,
+                    py::array_t<int> &status,
+                    py::array_t<double> &px,
+                    py::array_t<double> &py,
+                    py::array_t<double> &pz,
+                    py::array_t<double> &energy,
+                    py::array_t<double> &mass) 
+{
+    const double spare_mem = 1.5L;
+    double fill_size = pid.size();
+
+    // Reserve before appending for efficiency
+    auto &evt = private_access::member<Event_entry>(event);
+    evt.reserve(static_cast<int>(spare_mem*fill_size));
+    event.reset();
+
+    // Get a raw reference to numpy array
+    auto pid_ = pid.unchecked<1>();
+    auto status_ = status.unchecked<1>();
+    auto px_ = px.unchecked<1>();
+    auto py_ = py.unchecked<1>();
+    auto pz_ = pz.unchecked<1>();
+    auto energy_ = energy.unchecked<1>();
+    auto mass_ = mass.unchecked<1>();
+
+    for (int i = 0; i != fill_size; ++i) {
+        event.append(pid_[i], status_[i], 0, 0,
+                      px_[i], py_[i], pz_[i], energy_[i], mass_[i]);
+    }    
+}
+
+// Special case:
+// refills "event" stack with particles 
+// having pid and energy
+void fill_event(Event &event,
+                    const ParticleData &pd,
+                    py::array_t<int> &pid,
+                    py::array_t<double> &energy) 
+{
+    const double spare_mem = 1.5L;
+    double fill_size = pid.size();
+
+    // Reserve before appending for efficiency
+    auto &evt = private_access::member<Event_entry>(event);
+    evt.reserve(static_cast<int>(spare_mem*fill_size));
+    event.reset();
+
+    // Get a raw reference to numpy array
+    auto pid_ = pid.unchecked<1>();
+    auto energy_ = energy.unchecked<1>();
+
+    int id = 0;
+    double en = 0.;
+    double mass = 0;
+    double pz = 0.;
+
+    for (int i = 0; i != fill_size; ++i) {
+        id = pid_[i];
+        en = energy_[i];
+        mass = pd.findParticle(id)->m0();
+        pz = std::sqrt((en - mass)*(en + mass));    
+        event.append(pid_[i], 91, 0, 0, 0., 0., pz, en, mass);
+    }    
+}
+
```

Comment:
Is this special case really necessary? We keep the C++ bindings minimal, because these can break as we are using private details of Pythia that are not guaranteed to be stable by any convention.

I would prefer to use the generic fill_event and compute pz with numpy arrays. Might even be faster.

Again, the mass should come from the particle itself via the generated mass and not from ParticleData. For long-lived particles, there is not much difference, but for short-lived particles, using the generated_mass is important to maintain energy conservation.

Reply:
My case: I have only pid, energy arrays. Each 1e4-1e6 long. That is why I need to put them efficiently. So I don't have px, py, pz, or mass. I can use numpy for mass and pz calculations. The only thing which I find not nice: I need initiate long px and py zero arrays.

Anyway, I guess it is not such a big bottleneck. So, I removed the special case function.